pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -704,7 +709,76 @@ func (c *Controller) InstancesByPort(svc *model.Service, reqSvcPort int,
 		return inScopeInstances, nil
 	}
 
-	return c.endpoints.InstancesByPort(c, svc, reqSvcPort, labelsList)
+	outInstances, err := c.endpoints.InstancesByPort(c, svc, reqSvcPort, labelsList)
+	foreignInstances, err2 := c.getForeignServiceInstancesByPort(svc, reqSvcPort)
+
+	if err2 != nil {
+		err = multierror.Append(err, err2)
+	}
+
+	return append(outInstances, foreignInstances...), err
+}
+
+func (c *Controller) getForeignServiceInstancesByPort(svc *model.Service, reqSvcPort int) ([]*model.ServiceInstance, error) {
+	// Run through all the foreign instances, select ones that match these labels
+	// only if this is a kubernetes internal service and of ClientSideLB (eds) type
+	// as InstancesByPort is called by the aggregate controller. We dont want to include
+	// foreign instances for any other registry
+	if svc.Attributes.ServiceRegistry != string(serviceregistry.Kubernetes) ||
+		svc.MeshExternal || svc.Resolution != model.ClientSideLB {
+		return nil, nil
+	}
+
+	item, exists, err := c.services.GetStore().GetByKey(kube.KeyFunc(svc.Attributes.Name, svc.Attributes.Namespace))
+	if err != nil {
+		log.Infof("get foreignServiceInstancesByPort(%s, %s) => error %v", svc.Attributes.Name, svc.Attributes.Namespace, err)
+		return nil, err
+	}
+
+	if !exists {
+		return nil, nil
+	}
+
+	k8sService := item.(*v1.Service)
+	if k8sService.Spec.Selector == nil {
+		// this shouldn't be possible as we already check for resolution above. But just in case
+		return nil, nil
+	}
+	selector := klabels.Set(k8sService.Spec.Selector).AsSelectorPreValidated()
+
+	// Get the service port name so that we can construct the service instance
+	var servicePort *model.Port
+	for _, p := range svc.Ports {
+		if p.Port == reqSvcPort {
+			servicePort = p
+			break
+		}
+	}
+
+	out := make([]*model.ServiceInstance, 0)
+
+	c.RLock()
+	for _, fi := range c.foreignRegistryInstancesByIP {
```

Comment:
NBD, assuming the `foreignRegistryInstancesByIP` does not contain a huge list - but would it simplify if we have `foreignRegistryInstances` keyed by `instancesKey` (hostname and namespace) so that we can directly lookup and also reduce locking?

Reply:
So the foreign service instance is mostly a farce! What is really being shipped from one controller to another is the Endpoint. But costin did not want to add yet another interface 🤷 . So I had to reuse the appendInstanceHandlers, where the service instances have no real service (its just an empty object but with valid namespace attribute), and a valid endpoint. Ergo, the only unique thing about this instance is the IP.

Reply:
Good, this is a hotpoint but with low efficiency.